src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,436 @@
+using Magnum::Math::dot;
+
+namespace osp
+{
+
+    /**
+     * @brief Wrap an angle to the range [0, 2π), useful for keplerian orbits
+     */
+    inline constexpr double wrap_angle(double angle)
+    {
+        double wrapped = std::fmod(angle, 2.0 * M_PI);
+        if (wrapped < 0)
+        {
+            wrapped += 2.0 * M_PI;
+        }
+        return wrapped;
+    }
+
+    double stumpff_2(double z)
+    {
+        if (z > 0)
+        {
+            return (1 - std::cos(std::sqrt(z))) / z;
+        }
+        else if (z < 0)
+        {
+            return (std::cosh(std::sqrt(-z)) - 1) / -z;
+        }
+        else
+        {
+            return 1.0 / 2.0;
+        }
+    }
+
+    double stumpff_3(double z)
+    {
+        if (z > 0)
+        {
+            return (std::sqrt(z) - std::sin(std::sqrt(z))) / std::pow(z, 1.5);
+        }
+        else if (z < 0)
+        {
+            return (std::sinh(std::sqrt(-z)) - std::sqrt(-z)) / std::pow(-z, 1.5);
+        }
+        else
+        {
+            return 1.0 / 6.0;
+        }
+    }
+
+    double universal_kepler(double chi, double r_0, double v_0, double alpha, double delta_t, double mu)
+    {
+        double z = alpha * chi * chi;
+        double firstTerm = r_0 * v_0 / std::sqrt(mu) * chi * chi * stumpff_2(z);
+        double secondTerm = (1 - alpha * r_0) * chi * chi * chi * stumpff_3(z);
+        double thirdTerm = r_0 * chi;
+        double fourthTerm = std::sqrt(mu) * delta_t;
+        return firstTerm + secondTerm + thirdTerm - fourthTerm;
+    }
+
+    double d_universal_kepler_d_chi(double chi, double r_0, double v_0, double alpha, double delta_t, double mu)
+    {
+        double z = alpha * chi * chi;
+        double firstTerm = r_0 * v_0 / std::sqrt(mu) * chi * (1 - z * stumpff_3(z));
+        double secondTerm = (1.0 - alpha - r_0) * chi * chi * stumpff_2(z);
+        double thirdTerm = r_0;
+        return firstTerm + secondTerm + thirdTerm;
+    }
+
+    double solve_kepler_universal(double r_0, double v_0, double semiMajorAxis, double delta_t, double mu, int steps)
+    {
+        // https://orbital-mechanics.space/time-since-periapsis-and-keplers-equation/universal-variables.html
+        // This seems to converge disastrously slowly
+        // The author recommends using Laguerre's method for faster convergence
+        double alpha = 1.0 / semiMajorAxis;
+        double chi = std::sqrt(mu) * std::abs(alpha) * delta_t;
+        double delta;
+        do
+        {
+            double f_chi = universal_kepler(chi, r_0, v_0, alpha, delta_t, mu);
+            double df_chi = d_universal_kepler_d_chi(chi, r_0, v_0, alpha, delta_t, mu);
+            delta = f_chi / df_chi;
+            chi = chi - delta;
+            --steps;
+        } while (std::abs(delta) > 1.0E-10 && steps > 1);
+        return chi;
+    }
+
+    double solve_kepler_elliptic(double meanAnomaly, double eccentricity, int steps)
+    {
+        const double tolerance = 1.0E-10;
```

Comment:
Project's convention is to put const on the right side after the type. I usually put const wherever possible (literally everywhere, it's a little noisy though, if only const is the default)